astropy/table/table.py

```diff
@@ -4258,82 +4244,87 @@ def from_pandas(cls, dataframe, index=False, units=None):
-            masks.insert(0, np.zeros(len(dataframe), dtype=bool))
-
-        if units is None:
-            units = [None] * len(names)
-        else:
-            if not isinstance(units, Mapping):
-                raise TypeError('Expected a Mapping "column-name" -> "unit"')
-
-            not_found = set(units.keys()) - set(names)
-            if not_found:
-                warnings.warn(f"`units` contains additional columns: {not_found}")
-
-            units = [units.get(name) for name in names]
-
-        for name, column, data, mask, unit in zip(names, columns, datas, masks, units):
-            if column.dtype.kind in ["u", "i", "b"] and np.any(mask):
-                # Special-case support for pandas nullable int and bool
-                np_dtype = column.dtype.numpy_dtype
-                data = np.zeros(shape=column.shape, dtype=np_dtype)
-                data[~mask] = column[~mask]
-                out[name] = MaskedColumn(
-                    data=data, name=name, mask=mask, unit=unit, copy=False
-                )
-                continue
-
-            if data.dtype.kind == "O":
-                # If all elements of an object array are string-like or np.nan
-                # then coerce back to a native numpy str/unicode array.
-                string_types = (str, bytes)
-                nan = np.nan
-                if all(isinstance(x, string_types) or x is nan for x in data):
-                    # Force any missing (null) values to b''.  Numpy will
-                    # upcast to str/unicode as needed. We go via a list to
-                    # avoid replacing objects in a view of the pandas array and
-                    # to ensure numpy initializes to string or bytes correctly.
-                    data = np.array([b"" if m else d for (d, m) in zip(data, mask)])
-
-            # Numpy datetime64
-            if data.dtype.kind == "M":
-                from astropy.time import Time
-
-                out[name] = Time(data, format="datetime64")
-                if np.any(mask):
-                    out[name][mask] = np.ma.masked
-                out[name].format = "isot"
-
-            # Numpy timedelta64
-            elif data.dtype.kind == "m":
-                from astropy.time import TimeDelta
-
-                data_sec = data.astype("timedelta64[ns]").astype(np.float64) / 1e9
-                out[name] = TimeDelta(data_sec, format="sec")
-                if np.any(mask):
-                    out[name][mask] = np.ma.masked
+        from .dataframes import from_pandas
 
-            else:
-                if np.any(mask):
-                    out[name] = MaskedColumn(data=data, name=name, mask=mask, unit=unit)
-                else:
-                    out[name] = Column(data=data, name=name, unit=unit)
+        return from_pandas(dataframe, index=index, units=units)
+
+    @classmethod
+    def from_df(
+        cls, df: Any, index: bool = False, units: Mapping[str, Any] | None = None
```

Comment:
`df` isn't an argument name I'd be happy to see used as keyword (while clear in this context, it might be much less so in others). `index`being a boolean flag on the other hand, shouldn't be allowed as positional
```suggestion
        cls, df: Any, /, *, index: bool = False, units: Mapping[str, Any] | None = None
```